quil-rs/src/expression/simplification/by_hand.rs

```diff
@@ -311,156 +289,120 @@ fn simplify_infix(l: &Expression, op: InfixOperator, r: &Expression, limit: u64)
             // Addition with negation
             // a + (-b) = (-b) + a = a - b
             (
-                ref other,
+                _,
                 InfixOperator::Plus,
                 Expression::Prefix(PrefixExpression {
                     operator: PrefixOperator::Minus,
-                    ref expression,
+                    expression,
                 }),
-            )
-            | (
+            ) => self.simplify(interned::sub(left, expression.clone()), limit - 1),
+
+            (
                 Expression::Prefix(PrefixExpression {
                     operator: PrefixOperator::Minus,
-                    ref expression,
+                    expression,
                 }),
                 InfixOperator::Plus,
-                ref other,
-            ) => simplify_infix(other, InfixOperator::Minus, expression, limit - 1),
+                _,
+            ) => self.simplify(interned::sub(right, expression.clone()), limit - 1),
 
             // Subtraction with negation
 
             // a - (-b) = a + b
             (
-                ref left,
+                _,
                 InfixOperator::Minus,
                 Expression::Prefix(PrefixExpression {
                     operator: PrefixOperator::Minus,
-                    ref expression,
+                    expression,
                 }),
-            ) => simplify_infix(left, InfixOperator::Plus, expression, limit - 1),
+            ) => self.simplify(interned::add(left, expression.clone()), limit - 1),
 
             // -expression - right = smaller of [(-expression) - right, -(expression + right)]
             (
-                ref left @ Expression::Prefix(PrefixExpression {
+                Expression::Prefix(PrefixExpression {
                     operator: PrefixOperator::Minus,
-                    ref expression,
+                    expression,
                 }),
                 InfixOperator::Minus,
-                ref right,
+                _,
             ) => {
-                let original = sub!(left.clone(), right.clone());
-                let new = simplify_prefix(
-                    PrefixOperator::Minus,
-                    &simplify_infix(expression, InfixOperator::Plus, right, limit - 1),
-                    limit - 1,
-                );
-                min_by_key(original, new, size)
-            }
+                // Recreate the original structure but having simplified the contents
+                let original_structure = interned::sub(left.clone(), right.clone());
 
-            // Multiplication with negation
+                // Factor out the negation
+                let inner_sum = self.simplify(interned::add(expression.clone(), right), limit - 1);
+                let outer_neg = self.simplify(interned::neg(inner_sum), limit - 1);
```

Comment:
why does `self.simplify` need to be called again for `outer_neg`, isn't `inner_sum` already simplified?

Reply:
I guess I see that it would be fairly cheap to do, since `inner_sum` would already be cached, so you only need to do the `neg`

Reply:
Suppose `inner_sum` simplifies to `-x`; then you want `outer_neg` to be `x` instead of `-(-x)`.  Simplification is not preserved by constructors!  (That's I suppose a short summary of why this file is so nasty, simplification is a weird and picky property.)